metafacture-json/src/main/java/org/metafacture/json/JsonValidator.java

```diff
@@ -0,0 +1,194 @@
+        "Write valid and/or invalid output to locations specified with `writeValid` and `writeInvalid`. " +
+        "Set the JSON key for the record ID value with `idKey` (for logging output, defaults to `id`).")
+@In(String.class)
+@Out(String.class)
+@FluxCommand("validate-json")
+public final class JsonValidator extends DefaultObjectPipe<String, ObjectReceiver<String>> {
+
+    private static final Logger LOG = LoggerFactory.getLogger(JsonValidator.class);
+    private static final String DEFAULT_ID_KEY = "id";
+    private Schema schema;
+    private long fail;
+    private long success;
+    private FileWriter writeInvalid;
+    private FileWriter writeValid;
+    private String idKey = DEFAULT_ID_KEY;
+
+    /**
+     * @param url The URL of the schema to validate against.
+     */
+    public JsonValidator(final String url) {
+        initSchema(url);
+    }
+
+    /**
+     * @param writeValid The location to write valid data to.
+     */
+    public void setWriteValid(final String writeValid) {
+        this.writeValid = fileWriter(writeValid);
+    }
+
+    /**
+     * @param writeInvalid The location to write invalid data to.
+     */
+    public void setWriteInvalid(final String writeInvalid) {
+        this.writeInvalid = fileWriter(writeInvalid);
+    }
+
+    /**
+     * @param idKey The JSON key for the record ID value.
+     */
+    public void setIdKey(final String idKey) {
+        this.idKey = idKey;
+    }
+
+    @Override
+    public void process(final String json) {
+        try {
+            validate(json, new JSONObject(json) /* throws JSONException on syntax error */);
+        }
+        catch (final JSONException e) {
+            handleInvalid(json, null, e.getMessage());
+        }
+    }
+
+    private void validate(final String json, final JSONObject object) {
+        try {
+            schema.validate(object); // throws ValidationException if invalid
+            getReceiver().process(json);
+            ++success;
+            write(json, writeValid);
+        }
+        catch (final ValidationException e) {
+            handleInvalid(json, object, e.getAllMessages().toString());
+        }
+    }
+
+    @Override
+    protected void onCloseStream() {
+        close(writeInvalid);
+        close(writeValid);
+        LOG.debug("Success: {}, Fail: {}", success, fail);
+        super.onCloseStream();
+    }
+
+    private void initSchema(final String schemaUrl) {
+        if (schema != null) {
+            return;
+        }
+        SchemaLoaderBuilder schemaLoader = SchemaLoader.builder();
+        try {
+            final URL url = new URL(schemaUrl);
+            schemaLoader = schemaLoader.schemaJson(jsonFrom(url.openStream()))
+                    .resolutionScope(baseFor(url.toString()));
+        }
+        catch (final IOException e) {
+            LOG.info("Could not read as URL: {}, trying to load from class path", schemaUrl);
```

Comment:
Should this really be logged at `INFO` level? I'd suggest `DEBUG` instead.

Reply:
Yes, right, that would be confusing INFO output if we're intentionally using classpath resources.